### PR TITLE
feat: `DataFrame` supports unnesting multiple columns

### PR DESCRIPTION
**Pull request by jonahgao**
## Which issue does this PR close?
N/A

## Rationale for this change
#10044 has enabled SQL to support unnesting multiple columns, this PR adds the same functionality to `DataFrame`.

## Are these changes tested?
Yes

## Are there any user-facing changes?
Yes
Mark the `unnest_column` API as deprecated, and add a new API called `unnest_columns`.

### CODE REVIEW

File: datafusion/sqllogictest/test_files/unnest.slt [hunk 1]
```diff
@@ -383,5 +383,23 @@ select unnest(array_remove(column1, 3)) - 1 as c1, column3 from unnest_table;
...
 11 NULL
 
+# Unnest with non-nullable lists
+statement ok
+create table non_nullable_list(
+    c1 int[] not null,
+    c2 int[] not null
+) as values ([1,2,3], [4]);
+
+# Unnesting may produce NULLs even if the lists are non-nullable
```
Comment (jonahgao) [inline datafusion/sqllogictest/test_files/unnest.slt:393]: Added the test suggested by @jayzhan211 in https://github.com/apache/arrow-datafusion/pull/10044#discussion_r1563882878

File: datafusion/core/tests/dataframe/mod.rs [hunk 2]
```diff
@@ -1437,6 +1438,91 @@ async fn unnest_analyze_metrics() -> Result<()> {
...
+        "+------+------------+------------+--------+",
+    ];
+    assert_batches_sorted_eq!(expected, &results);
+
+    Ok(())
+}
+
+/// Test unnesting a non-nullable list.
+#[tokio::test]
+async fn unnest_non_nullable_list() -> Result<()> {
```
Comment (jonahgao) [inline datafusion/core/tests/dataframe/mod.rs:1498]: Added the test suggested by @jayzhan211 in https://github.com/apache/arrow-datafusion/pull/10044#discussion_r1563882878

**Review by jayzhan211 — APPROVED** [review]
👍

File: datafusion/core/tests/dataframe/mod.rs [hunk 3]
```diff
@@ -1437,6 +1438,91 @@ async fn unnest_analyze_metrics() -> Result<()> {
...
+        "|      |            | 4          | c      |",
+        "|      | 2.2        | 1          | b      |",
+        "|      | 3.3        | 2          | b      |",
+        "|      | 4.4        |            | b      |",
+        "| 1    |            |            | a      |",
+        "| 2    | 1.1        |            | a      |",
+        "| 3    |            |            | a      |",
+        "+------+------------+------------+--------+",
+    ];
+    assert_batches_sorted_eq!(expected, &results);
```
Comment (jayzhan211) [inline datafusion/core/tests/dataframe/mod.rs:1466]: I think the result is deterministic so checking with sorted is not necessary.

Comment (jayzhan211) [inline datafusion/core/tests/dataframe/mod.rs:1466]: I think the result is deterministic, so we can check without sorting

Then, we can have a more straightforward result
```
[
    "+------+------------+------------+--------+",
    "| list | large_list | fixed_list | string |",
    "+------+------------+------------+--------+",
    "| 1    |            |            | a      |",
    "| 2    | 1.1        |            | a      |",
    "| 3    |            |            | a      |",
    "|      | 2.2        | 1          | b      |",
    "|      | 3.3        | 2          | b      |",
    "|      | 4.4        |            | b      |",
    "|      |            | 3          | c      |",
    "|      |            | 4          | c      |",
    "|      |            |            | d      |",
    "+------+------------+------------+--------+",
]
```

File: datafusion/core/tests/dataframe/mod.rs [hunk 4]
```diff
@@ -1437,6 +1438,91 @@ async fn unnest_analyze_metrics() -> Result<()> {
...
```
Comment (jayzhan211) [inline datafusion/core/tests/dataframe/mod.rs:1491]: Same here

Comment (jonahgao) [reply datafusion/core/tests/dataframe/mod.rs]: Done.

File: datafusion/core/tests/dataframe/mod.rs (hunk 3 above)
Comment (jonahgao) [reply datafusion/core/tests/dataframe/mod.rs]: > I think the result is deterministic, so we can check without sorting

Updated. Thank you @jayzhan211 . It looks much better now.

**Comment by jayzhan211** [comment]
Nice! Thanks @jonahgao

**Pull request merged**